wire-ios-data-model/Tests/Source/ManagedObjectContext/DatabaseMigrationTests+UserClientUniqueness.swift

```diff
@@ -0,0 +1,204 @@
+//
+// Wire
+// Copyright (C) 2023 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import XCTest
+import Foundation
+@testable import WireDataModel
+
+final class DatabaseMigrationTests_UserClientUniqueness: DatabaseBaseTest {
+
+    typealias MigrationAction = (NSManagedObjectContext) throws -> Void
+
+    private let bundle = Bundle(for: ZMManagedObject.self)
+    private let clientID = "abc123"
+    private let tmpStoreURL = URL(fileURLWithPath: "\(NSTemporaryDirectory())databasetest/")
+    private let dataModelName = "zmessaging"
+
+    func testMigratingToMessagingStore_2_106_PreventsDuplicateUserClients() throws {
+        try migrateStore(
+            sourceVersion: "2.105.0",
+            destinationVersion: "2.106.0",
+            preMigrationAction: { context in
+                // insert some duplicates
+                try insertDuplicateClients(with: clientID, in: context)
+                let clients = try fetchClients(with: clientID, in: context)
+                XCTAssertEqual(clients.count, 2)
+            },
+            postMigrationAction: { context in
+                var clients: [UserClient]
+
+                // verify it deleted duplicates
+                clients = try fetchClients(with: clientID, in: context)
+                XCTAssertEqual(clients.count, 1)
+
+                // verify we can't insert duplicates
+                context.mergePolicy = NSMergePolicy.mergeByPropertyObjectTrump
+                try insertDuplicateClients(with: clientID, in: context)
+                clients = try fetchClients(with: clientID, in: context)
+                XCTAssertEqual(clients.count, 1)
+            }
+        )
+    }
+
+    // MARK: - Migration Helpers
+
+    private func migrateStore(
+        sourceVersion: String,
+        destinationVersion: String,
+        preMigrationAction: MigrationAction,
+        postMigrationAction: MigrationAction
+    ) throws {
```

Comment:
Very nice 👍🏻

Reply:
It may also be nice to factor this out into a superclass so that we can reuse it for future migrations. We could of course do that later when we need to.